### PR TITLE
Fix inference with ref parameter with maybe-null value

### PR DESCRIPTION
**Pull request by jcouv**
Previously, the L-value of a `ref` argument would contribute its type+nullability. Now we're using the R-value.

Fixes https://github.com/dotnet/roslyn/issues/47663
Fixes https://github.com/dotnet/roslyn/issues/35534

Note that the inference in some scenario becomes a bit worse (for example with `CompareExchange`) when there is a `ref` argument and `null`. There is a [known issue](https://github.com/dotnet/roslyn/issues/43536) with `null` not contributing nullability to method type inference.

We could discuss whether to hold off this change until the issue with `null` is fixed. 
I think the current change can be taken as-is, because (1) they involve having a nullable but not-null `ref`, which is uncommon for `CompareExchange`, and (2) they can be mitigated with explicit type parameters.

### CODE REVIEW

**Comment by jcouv** [comment]
Tagging @stephentoub @jkotas. 
This PR changes the way `ref` arguments contribute to method type inference with nullability. Instead of using the declared type of the ref (L-value), we're using it's tracked value (R-value).

But, due to another open issue with `null` literals not contributing nullness to the inference, this makes the inference worse in some scenarios. 

For example, calls to `CompareExchange` when the arguments are either thought be not-null or the `null` literal.
The tests show the 3 impacted scenarios for `CompareExchange`:
- `[InlineData("locationNonNull", "valueNonNull", "null", true)]  `
- `[InlineData("locationNonNull", "null", "comparandNonNull", false)]`
- `[InlineData("locationNonNull", "null", "null", false)]`

I believe those are uncommon (the location argument would typically be maybe-null).

Any other generic API call with a `ref` and a `null` literal could be impacted in the same way. We were unable to fix the problem with `null` in VS 16.8. Aiming for 16.9 at this point.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -29411,48 +29411,206 @@ public void M(string? s)
...
+        public void RefParameter_Issue_47663()
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class C
+{
+    static void X1<T>([NotNullIfNotNull(""value"")] ref T location, T value) where T : class? { }
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29485]: How did your change affect this scenario? I can get how before we inferred a `double[]` type argument and now we infer a `double[]?` type argument, but how did that result in the postcondition for the NotNullIfNotNullAttribute starting to be respected?

**Comment by stephentoub** [comment]
@jcouv, have you tried building dotnet/runtime?  Does this introduce any new warnings-as-error as part of its build? Thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -48735,53 +48893,53 @@ void Test1(CL1 x1, CL1? z1)
     {
         var y1 = new { p1 = x1, p2 = z1 };
         x1 = y1.p1 ?? x1;
-        x1 = y1.p2;
+        x1 = y1.p2; // 1
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48896]: Why did this change, don't see any `ref` usage here.?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -125367,16 +125522,13 @@ void M(object? o1)
         // 2 states * 3 states * 3 states = 18 combinations
         [InlineData("locationNonNull", "valueNonNull", "comparandNonNull", true)]
         [InlineData("locationNonNull", "valueNonNull", "comparandMaybeNull", true)]
-        [InlineData("locationNonNull", "valueNonNull", "null", true)]
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:125370]: Why was this case removed? This reads to me as: "if the location contains null, write non-null to it."

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -125389,14 +125541,110 @@ public void CompareExchange_LocationNullState(string location, string value, str
...
+                    // (12,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(12, 13)
+                });
+        }
+
+        // This test should be merged back into CompareExchange_LocationNullState when the issue with inference with null is resolved
+        // Currently differs by a diagnostic
+        [Theory]
+        [InlineData("locationNonNull", "null", "comparandNonNull", false)]
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:125634]: Should we also add `[InlineData("locationNonNull", "null", "null", false)]` here?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -29411,48 +29411,206 @@ public void M(string? s)
...
+                //     C()
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "f2").WithLocation(14, 5),
+                // (18,16): warning CS8601: Possible null reference assignment.
+                //         X2(ref f2, bar);
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "f2").WithLocation(18, 16)
+                );
+        }
+
+        [Fact]
+        public void RefParameter_MiscPermutations()
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29511]: Correct. None of those tests change the state of the variables so L-value == R-value. I just wanted to check how multiple `ref` arguments were combined.

---
In reply to: [493889166](https://github.com/dotnet/roslyn/pull/47952#discussion_r493889166) [](ancestors = 493889166)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:125370]: Those three scenario were moved to their own test below (for now). Currently `null` literals do not contribute their nullability to method type re-inference, so the diagnostics are different (ie. worse).

---
In reply to: [493903400](https://github.com/dotnet/roslyn/pull/47952#discussion_r493903400) [](ancestors = 493903400)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:125370]: Sorry, I must not have noticed it below.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48896]: Correct. I just added the `// N` comments for clarity. One diagnostic was removed in `Test10` (which does involve `ref`), but that's all.

---
In reply to: [493894972](https://github.com/dotnet/roslyn/pull/47952#discussion_r493894972) [](ancestors = 493894972)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29485]: We used to produce a warning on `ref f1` for assigning maybe-null value inbound to the first parameter (which was inferred as `double[]!`.
For `ref f2` we now produce a diagnostic on the outbound assignment from parameter `double[]?` to local typed `double[]!`.

---
In reply to: [493893383](https://github.com/dotnet/roslyn/pull/47952#discussion_r493893383) [](ancestors = 493893383)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29485]: Makes complete sense now, thanks!

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29511]: Sounds good, just wanted to verify.

**Review by jaredpar — APPROVED** [review]

**Comment by jcouv** [comment]
Testing this change in `runtime` repo caused impact in one location. Here's the PR to update `runtime` repo: https://github.com/dotnet/runtime/pull/42707

**Comment by jcouv** [comment]
The VS validation build has two new warnings because of this change:
```
Q:\cmd\1r\src\vc\projbld\ExternalBuildFramework\Impl\Services\BuildProjectScannerService.cs(1351,85): error CS8625: Cannot convert null literal to non-nullable reference type.
Q:\cmd\1r\src\vc\projbld\ExternalBuildFramework\Impl\Services\BuildProjectScannerService.RunningTaskMonitor.cs(78,72): error CS8625: Cannot convert null literal to non-nullable reference type.
```

[BuildProjectScannerService.cs](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS?path=%2Fsrc%2Fvc%2Fprojbld%2FExternalBuildFramework%2FImpl%2FServices%2FBuildProjectScannerService.cs&version=GBdev%2Fvslsnap%2Finsertions%2Fmain.20200915225523)
[BuildProjectScannerService.RunningTaskMonitor.cs](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS?path=%2Fsrc%2Fvc%2Fprojbld%2FExternalBuildFramework%2FImpl%2FServices%2FBuildProjectScannerService.RunningTaskMonitor.cs&version=GBdev%2Fvslsnap%2Finsertions%2Fmain.20200915225523)

Those are consequences of the known issue with `null` literals not contributing properly to method type inference. The recommended fix is to use `null!` for now.

FYI @dibarbet

I'll add a test to this PR to illustrate that scenario.

**Comment by dibarbet** [comment]
> The VS validation build has two new warnings because of this change:

@jcouv would you mind if I updated your validation PR (I assume https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/277093?_a=overview) to include the suggested fixes.  That way I can be sure cloud build is passing and I can easily cherry-pick the commit over to the proper insertion

**Comment by jcouv** [comment]
@dibarbet Thanks. I've pushed two commits to the [validation PR](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/277093) and will ping you once it is green.

**Comment by RikkiGibson** [comment]
[Here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABAZguYCYKBhCgb0YVhHdswAsFALIBGADwAVAHwAKBHABmFBRRgyk23ZwCUQkYIYirumRQC8RgNxnhAXxcUPbDpKmcVppbmHlbMMgAMAPwUUPYUAEQQ8c5B1tIyapoxBgTExilpwlAAdAoA9gDKMAgAlrgA5gGOFAD0LTEAFmVEKBQA7hAIuB7uqZ5j3hLSrAEeFoUcEdFQtg6JySEiYVEx3A65hAWFsplaKwZQJkdpK6WV1XWN+a3tUF09/YPDY6NWXmK+cSzMbzQrbZZxdbXawndRnAzxYDxfKbIp3Kq1BpNF6dbqEXq4MowT5DEaMUZAA=) are a few scenarios that we should try to make work in a future PR. I think `M2()` may not have desired behavior due to #43536-- it may warn on the `null` literal argument instead of on accessing members on the return value.

```cs
#nullable enable

public class C {
    public void M1<T>(ref T t1, T t2)
    {
        t1 = t2;
    }
    
    public void M2()
    {
        string? s = "a";
        M1(ref s, null);
        s.ToString(); // should warn
    }
    
    public void M3()
    {
        string? s1 = "a";
        string? s2 = null;
        M1(ref s1, s2);
        s1.ToString(); // should warn
    }
    
    public void M4()
    {
        string? s = "a";
        M1(ref s, "b");
        s.ToString(); // should not warn
    }
}
```

**Comment by dibarbet** [comment]
@jcouv forgot to mention - to get a passing cloudbuild you may need the two extra commits from this PR
https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/276948?_a=commits

**Comment by jcouv** [comment]
@dibarbet Thanks. I indeed had to add the two commits from that other PR to get the build to pass. 
Feel free to grab the commits from https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/277093 (there are five, each adding a suppression. They can be squashed).
We'll go ahead and merge the roslyn change.

**Pull request merged**